wrapper/source_hosts.py

```diff
@@ -0,0 +1,773 @@
+                                                   self._converter_out,
+                                                   update_source, volume_id))
+
+    def _export_volumes_from_converter(self):
+        """
+        SSH to source conversion host and start an NBD export. Start the UCI
+        with /dev/vdb, /dev/vdc, etc. attached, then pass JSON input to request
+        nbdkit exports from the V2V wrapper. Volume mapping step 4: fill in the
+        URL to the volume's matching NBD export.
+        """
+        logging.info('Exporting volumes from source conversion host...')
+
+        # Create a temporary directory on source conversion host
+        self.tmpdir = self._converter_out(['mktemp', '-d', '-t', 'v2v-XXXXXX'])
+        logging.info('Source conversion host temp dir: %s', self.tmpdir)
+
+        # Choose NBD ports for inside the container
+        port = 10809
+        port_map = {}          # Map device path to port number, for export_nbd
+        reverse_port_map = {}  # Map port number to device path, for forwarding
+        nbd_ports = []         # podman arguments for NBD ports
+        device_list = []       # podman arguments for block devices
+        for path, mapping in self.volume_map.items():
+            volume_id = mapping.source_id
+            dev_path = mapping.source_dev
+            uci_dev_path = mapping.source_dev+'-v2v'
+            logging.info('Exporting %s from volume %s', dev_path, volume_id)
+            nbd_ports.extend(['-p', '{0}'.format(port)])
+            device_list.extend(['--device', dev_path+':'+uci_dev_path])
+            reverse_port_map[port] = path
+            port_map[uci_dev_path] = port
+            port += 1
+
+        # Copy the port map as input to the source conversion host wrapper
+        self._converter_val(['mkdir', '-p', self.tmpdir+'/lib'])
+        self._converter_val(['mkdir', '-p', self.tmpdir+'/log'])
+        self._converter_val(['mkdir', '-p', self.tmpdir+'/input'])
+        ports = json.dumps({'nbd_export_only': port_map})
+        export_input = '/tmp/nbd_conversion.json'
+        with open(export_input, 'w+') as conversion:
+            conversion.write(ports)
+        self._converter_scp(export_input, self.tmpdir+'/input/conversion.json')
+
+        # Run UCI on source conversion host. Create a temporary directory to
+        # use as the UCI's /data directory so more than one can run at a time.
+        ssh_args = ['sudo', 'podman', 'run', '--detach']
+        ssh_args.extend(['--volume', '/var/tmp:/var/tmp:z'])
+        ssh_args.extend(['--volume', '/var/lock:/var/lock:z'])
+        ssh_args.extend(['--volume', self.tmpdir+':/data:z'])
+        ssh_args.extend(['--volume', self.tmpdir+'/lib:/var/lib/uci:z'])
+        ssh_args.extend(['--volume', self.tmpdir+'/log:/var/log/uci:z'])
+        ssh_args.extend(['--volume',
+                         '/opt/vmware-vix-disklib-distrib:'
+                         '/opt/vmware-vix-disklib-distrib'])
```

Comment:
Things like this are already in common so that we don't have some hard-coded paths in multiple places.

Reply:
Whoops, done.